File: CS/Behavioural/Command/Calculator.cs
namespace Pattern.Behavioural.Command
{
	internal static partial class CommandPatternRunner
	{
		internal class Calculator
		{
			public double CurrentValue {get; private set; }
			private Stack<ICommand> stack = new Stack<ICommand>();
			public Calculator() 
			{
				CurrentValue = 0;
			}

			public void Do(ICommand command)
			{
				CurrentValue = command.Do(CurrentValue);
				stack.Push(command);
			}

			public void Undo()
			{
				var lastCommand = stack.Pop();
				CurrentValue = lastCommand.Undo(CurrentValue);
			}

		}

    }
}

File: VB/Behavioural/Command/Calculator.vb
Friend Class Calculator

	Public Value As Double
	Private _commands As Stack(Of ICommand)

	Sub New()
		Value = 0
		_commands = New Stack(Of ICommand)
	End Sub

	Public Sub DoCommand(command As ICommand)
		Value = command.DoCommand(Value)
		_commands.Push(command)
	End Sub

	Public Sub UndoCommand()
		Dim command = _commands.Pop
		Value = command.UndoCommand(Value)
	End Sub

	Public Sub Reset()
		Value = 0
		_commands.Clear()
	End Sub

End Class
